File: checkpoint0/src/main/java/com/springernature/checkpoint0/Game.java
package com.springernature.checkpoint0;

import java.util.ArrayList;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static com.springernature.checkpoint0.Cell.*;
import static java.util.Arrays.asList;

class Game {
    private final Board board;
    private final Cell currentPlayer;

    Game() {
        this(new Board(IntStream.range(0, 3)
                .mapToObj(index -> new ArrayList<>(asList(Empty, Empty, Empty)))
                .collect(Collectors.toList())), X);
    }

    private Game(Board board, Cell cell) {
        this.board = board;
        this.currentPlayer = cell;
    }

    Game makeMove(Move move, Cell cell) throws InvalidMove {
        if (move.row < 0 || move.row >= board.size() || move.column < 0 || move.column >= board.size()) {
            throw new InvalidMove("Invalid move: (" + move.row + "," + move.column + ") it must be between 0 and " + (board.size() - 1));
        }

        return new Game(board.set(move.row, move.column, cell), Objects.equals(currentPlayer, X) ? O : X);
    }

    Board board() {
        return new Board(board);
    }

    Cell nextCellToPlace() {
        return currentPlayer;
    }

    boolean isGameOver() {
        // TODO implement later
        return false;
    }

    boolean isGameADraw() {
        // TODO implement later
        return false;
    }

    static class InvalidMove extends Exception {
        InvalidMove(String message) {
            super(message);
        }
    }
}

File: checkpoint1/src/main/kotlin/com/springernature/checkpoint1/Game.kt
package com.springernature.checkpoint1

import com.springernature.checkpoint1.Cell.*

class Game(private val board: Board = Board(List(3, { List(3, { Empty }) })),
           private val currentPlayer: Cell = X) {

    fun makeMove(move: Move, cell: Cell): Game {
        if (move.row < 0 || move.row >= board.size || move.column < 0 || move.column >= board.size) {
            throw InvalidMove("Invalid move: (" + move.row + "," + move.column + ") it must be between 0 and " + (board.size - 1))
        }

        return Game(board.set(move.row, move.column, cell), if (currentPlayer == X) O else X)
    }

    fun board() = Board(board)

    fun nextCellToPlace() = currentPlayer

    // TODO implement later
    val isGameOver: Boolean
        get() = false

    // TODO implement later
    val isGameADraw: Boolean
        get() = false

    // TODO implement later
    internal class InvalidMove(message: String) : Exception(message)
}
